### PR TITLE
Non-iterative solution for stability parameter in stable boundary layers

### PR DESCRIPTION
**Pull request by akshaysridhar**
## Purpose and Content
Following Gryanik et al. (2021) DOI: 10.1029/2021MS002590, we include a non-iterative solution for stability parameter ζ (and therefore Monin-Obukhov length) in cases where ζ >= 0. Design documentation has been updated. 

1. Neutral b.l. tolerance is specified at just one point now (via user interface function `surface_conditions`). Value is kept at cp_d/100 following PR#66.  
2. Keyword argument added to `surface_conditions` to allow user to select whether they want to use the non-iterative procedure specified above. 
3. For stability functions which are only for zeta >= 0, Businger-Dyer functions are applied as the default option in cases with ζ <= 0. 
4. Adds convergence checks for the `Gryanik` and `Grachev` type universal functions. 

## Benefits
Optional removal of dependence on user specified tolerances in ζ >= 0 solutions. Following Gryanik et al. (2021) the modifications are expected to be useful for a wider range of stability parameters ζ if the `Gryanik / Grachev` type stability functions are applied. 
 
## Risks
Effect on existing solutions. Can be mitigated by full integration testing prior to merge. Two changes to user interface via keyword arguments for `tol_neutral` and `noniterative_stable_sol`. 

## Linked Issues
#35 

cc @yairchn @szy21 

## PR Checklist
- [x] This PR has a corresponding issue OR is linked to an SDI.
- [x] I have followed CliMA's codebase [contribution](https://clima.github.io/ClimateMachine.jl/latest/Contributing/) and [style](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) guidelines OR N/A.
- [x] I have followed CliMA's [documentation policy](https://github.com/CliMA/policies/wiki/Documentation-Policy).
- [x] I have checked all issues and PRs and I certify that this PR does not duplicate an open PR.
- [x] I linted my code on my local machine prior to submission OR N/A.
- [x] Unit tests are included.
- [x] Code used in an integration test OR N/A.
- [x] All tests ran successfully on my local machine OR N/A.
- [x] All classes, modules, and function contain docstrings OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by szy21** [comment]
Could we keep the iterative solution and make the analytical solution a (default) option?

File: Project.toml [hunk 1]
```diff
@@ -4,6 +4,7 @@ authors = ["Climate Modeling Alliance"]
 version = "0.4.7"
 
 [deps]
+CLIMAParameters = "6eacf6c3-8458-43b9-ae03-caf5306d3d53"
```
Comment (charleskawczynski) [inline Project.toml:7]: This shouldn't be needed.

Comment (charleskawczynski) [inline Project.toml:7]:
```suggestion
```

**Review by charleskawczynski — CHANGES_REQUESTED** [review]
CLIMAParameters should not be apart of the `SurfaceFluxes.jl/Project.toml`

**Comment by akshaysridhar** [comment]
> Could we keep the iterative solution and make the analytical solution a (default) option?

Done as a keyword argument now. Convergence checks include the same sweep over states but with this keyword for noniterative solves enabled.

File: Project.toml (hunk 1 above)
Comment (akshaysridhar) [reply Project.toml]: Fixed

Comment (akshaysridhar) [reply Project.toml]: Fixed (removed from Project.toml)

**Review by charleskawczynski — APPROVED** [review]

**Comment by akshaysridhar** [comment]
TC Buildkite output: https://buildkite.com/clima/turbulenceconvection-ci/builds/11959#0183a3dd-9bbb-4092-9085-af107b6005e8

File: src/SurfaceFluxes.jl [hunk 2]
```diff
@@ -288,14 +289,16 @@ function surface_conditions(
...
+    noniterative_stable_sol::Bool = false,
 ) where {FT}
     uft = SFP.universal_func_type(param_set)
-    L_MO = obukhov_length(param_set, sc, uft, scheme; tol, maxiter, soltype)
+    L_MO = obukhov_length(param_set, sc, uft, scheme; tol, tol_neutral, maxiter, soltype, noniterative_stable_sol)
     ustar = compute_ustar(param_set, L_MO, sc, uft, scheme)
-    Cd = momentum_exchange_coefficient(param_set, L_MO, sc, uft, scheme)
-    Ch = heat_exchange_coefficient(param_set, L_MO, sc, uft, scheme)
+    Cd = momentum_exchange_coefficient(param_set, L_MO, sc, uft, scheme, tol_neutral)
+    Ch = heat_exchange_coefficient(param_set, L_MO, sc, uft, scheme, tol_neutral)
```
Comment (szy21) [inline src/SurfaceFluxes.jl:301]: why do we need `tol_neutral` in `momentum_exchange_coefficient` and `heat_exchange_coefficient`?

Comment (akshaysridhar) [reply src/SurfaceFluxes.jl:301]: The coefficients take one of two forms based on whether we have a neutral layer or not; so the idea was to allow the user to change that tolerance only in the `surface_conditions` interface, the rest should all follow from that one point

Comment (akshaysridhar) [reply src/SurfaceFluxes.jl:301]: e.g. line https://github.com/CliMA/SurfaceFluxes.jl/blob/162a8c8dc619f8cd65fcbbe814d1b66789cef910/src/SurfaceFluxes.jl#L608 for the heat exchange terms, line https://github.com/CliMA/SurfaceFluxes.jl/blob/162a8c8dc619f8cd65fcbbe814d1b66789cef910/src/SurfaceFluxes.jl#L565 for the momentum exchange terms

**Comment by akshaysridhar** [comment]
bors try

**Review by szy21 — APPROVED** [review]

**Pull request merged**